File: app/src/main/java/com/anadol/mindpalace/data/statistic/StatisticService.java
package com.anadol.mindpalace.data.statistic;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;

import com.anadol.mindpalace.data.group.DataBaseSchema;
import com.anadol.mindpalace.data.group.GroupCursorWrapper;

import java.util.ArrayList;
import java.util.List;

public class StatisticService {

    public static List<GroupStatisticItem> getGroupStatistic(Context context) {
        List<GroupStatisticItem> statisticItems = new ArrayList<>();

        ContentResolver contentResolver = context.getContentResolver();

        GroupCursorWrapper groupCursor = new GroupCursorWrapper(contentResolver.query(
                DataBaseSchema.Groups.CONTENT_URI,
                null, null, null, null));

        if (groupCursor.getCount() != 0) {
            groupCursor.moveToFirst();

            GroupStatisticItem item;
            String name;
            String uuidGroup;
            int needToLearn = 0;
            int learning = 0;
            int learned = 0;


            while (!groupCursor.isAfterLast()) {
                name = groupCursor.getString(groupCursor.getColumnIndex(DataBaseSchema.Groups.NAME_GROUP));
                uuidGroup = groupCursor.getString(groupCursor.getColumnIndex(DataBaseSchema.Groups.UUID));

                // Need to learn
                Cursor wordCursor = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI,
                        new String[]{"COUNT(" + DataBaseSchema.Words._ID + ") AS count"},
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND (" + DataBaseSchema.Words.TIME + " = ? OR " + DataBaseSchema.Words.TIME
                                + " IS NULL)",
                        new String[]{uuidGroup, "0"}, null);

                if (wordCursor != null) {
                    wordCursor.moveToFirst();
                    needToLearn = wordCursor.getInt(0);
                    wordCursor.close();
                }

                // Learning
                wordCursor = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI,
                        new String[]{"COUNT(" + DataBaseSchema.Words._ID + ") AS count"},
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND " + DataBaseSchema.Words.TIME + " != ? AND " + DataBaseSchema.Words.EXAM + " = ?",
                        new String[]{uuidGroup, "0", "0"}, null);

                if (wordCursor != null) {
                    wordCursor.moveToFirst();
                    learning = wordCursor.getInt(0);
                    wordCursor.close();
                }

                // Learned
                wordCursor = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI,
                        new String[]{"COUNT(" + DataBaseSchema.Words._ID + ") AS count"},
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND " + DataBaseSchema.Words.EXAM + " = ?",
                        new String[]{uuidGroup, "1"}, null);

                if (wordCursor != null) {
                    wordCursor.moveToFirst();
                    learned = wordCursor.getInt(0);
                    wordCursor.close();
                }

                item = new GroupStatisticItem(name, needToLearn, learning, learned);
                statisticItems.add(item);

                groupCursor.moveToNext();
            }
        }
        groupCursor.close();

        return statisticItems;
    }

}

File: app/src/main/java/com/anadol/mindpalace/data/statistic/StatisticService.kt
package com.anadol.mindpalace.data.statistic

import android.annotation.SuppressLint
import android.content.Context
import com.anadol.mindpalace.data.group.DataBaseSchema
import com.anadol.mindpalace.data.group.GroupCursorWrapper
import com.anadol.mindpalace.domain.utils.getFirstInt

object StatisticService {

    @SuppressLint("Recycle")
    fun getGroupStatistic(context: Context): List<GroupStatisticItem> {
        val statisticItems: MutableList<GroupStatisticItem> = ArrayList()
        val contentResolver = context.contentResolver

        GroupCursorWrapper(
            contentResolver.query(DataBaseSchema.Groups.CONTENT_URI, null, null, null, null)
        ).use { groupCursor ->
            if (groupCursor.count != 0) {
                groupCursor.moveToFirst()

                var item: GroupStatisticItem
                var name: String?
                var uuidGroup: String

                while (!groupCursor.isAfterLast) {
                    name = groupCursor.getString(groupCursor.getColumnIndex(DataBaseSchema.Groups.NAME_GROUP))
                    uuidGroup = groupCursor.getString(groupCursor.getColumnIndex(DataBaseSchema.Groups.UUID))

                    // Need to learn
                    val needToLearn = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI, arrayOf("COUNT(" + DataBaseSchema.Words._ID + ") AS count"),
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND (" + DataBaseSchema.Words.TIME + " = ? OR " + DataBaseSchema.Words.TIME
                                + " IS NULL)", arrayOf(uuidGroup, "0"), null
                    )?.getFirstInt() ?: 0

                    // Learning
                    val learning = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI,
                        arrayOf("COUNT(" + DataBaseSchema.Words._ID + ") AS count"),
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND " + DataBaseSchema.Words.TIME + " != ? AND " + DataBaseSchema.Words.EXAM + " = ?",
                        arrayOf(uuidGroup, "0", "0"),
                        null
                    )?.getFirstInt() ?: 0

                    // Learned
                    val learned = contentResolver.query(
                        DataBaseSchema.Words.CONTENT_URI,
                        arrayOf("COUNT(" + DataBaseSchema.Words._ID + ") AS count"),
                        DataBaseSchema.Words.UUID_GROUP + " = ? AND " + DataBaseSchema.Words.EXAM + " = ?",
                        arrayOf(uuidGroup, "1"),
                        null
                    )?.getFirstInt() ?: 0

                    item = GroupStatisticItem(name, needToLearn, learning, learned)
                    statisticItems.add(item)
                    groupCursor.moveToNext()
                }
            }
        }

        return statisticItems
    }

}
